Isis UX: hide/show toolbar buttons

dgrammatiko opened the pull request:
#### Simplify the backend

This is a redo of @phproberto ’s #2111 with the addition for the button unfeature
##### What’s changed?

Most of the buttons (on a list view) are useless until the user actually select some items. This PR prevents hides those buttons till something is selected!
#### Preview

Nothing selected:
![screenshot 2015-07-30 18 19 15](https://cloud.githubusercontent.com/assets/3889375/8987150/451d2f2c-36e8-11e5-9eec-47dd94ffc5d6.png)

Item selected:
<img width="770" alt="screenshot 2015-07-30 18 19 39" src="https://cloud.githubusercontent.com/assets/3889375/8987166/559e8544-36e8-11e5-8a5f-c82d5e419477.png">

dgrammatiko committed:
hide/show toolbar buttons

brianteeman commented:
LOVE IT :heart: Thanks for resurrecting this.

This makes everything so much cleaner :+1:

Also note that if you have more than one item selected then the edit button is also correctly removed which also resolves #7589 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7592">issues.joomla.org/joomla-cms/7592</a>.</sub>

photodude commented:
:100:

justinherrin commented:
Perhaps this is intentional, but the "Edit" button does not re-appear when 1 category is left selected after 2 categories were previously selected. To better explain...

1. Have **only 2** categories in your Category Manager
2. Select "Category A" checkbox... all the buttons appear
3. With "Category A" still selected, also select "Category B"... the "Edit" button disappears
4. Unselect either of the categories (so now you just have 1 category selected again)... the "Edit" button does not re-appear

When tested with **more than 2** categories, the "Edit" button re-appears just fine :)

brianteeman commented:
I can confirm what @justinherring reports<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7592">issues.joomla.org/joomla-cms/7592</a>.</sub>

brianteeman commented:
oops sorry @justinherrin <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7592">issues.joomla.org/joomla-cms/7592</a>.</sub>

dgrammatiko committed:
edit show/hide

dgt41 commented:
One minor change to get edit button hide/appear properly. Please keep in mind that you should clear the browser’s cache!

brianteeman commented:
Seems ok now

justinherrin commented:
Looks great now!

uthorat commented:
Yes It looks great. 
Thank you @dgt41

infograf768 commented:
My comment on this will be the same as in the original proposition:
-----------
The code looks so nice to us that we do not see It defeats our purpose, i.e. make the UI easier.

In fact, we, experienced users who test this, do indeed know the possibilities of action will display when we tick an item, but a newbie will never know and I imagine that even not so newbies will hesitate...

The toolbar not only displays icons but these icons have a legend which explain their purpose. How would a newbie (client) know that one or multiple items may be trashed, archived, checked-in, part of a batch, etc. ?
People NEED to know what is available... It is essential for a good user experience.

There was a proposal to dim the buttons. That would solve the issue.
Another solution would be to create a new parameter in the template, and disable the hiding by default.

brianteeman commented:
I am not opposed to you creating a parameter with the default being ON
(hide)

Time moves on and Joomla has to move on as well. This is a very common UI
pattern now.

On 31 July 2015 at 09:05, infograf768 <notifications@github.com> wrote:

> My comment on this will be the same as in the original proposition:
>
> The code looks so nice to us that we do not see It defeats our purpose,
> i.e. make the UI easier.
>
> In fact, we, experienced users who test this, do indeed know the
> possibilities of action will display when we tick an item, but a newbie
> will never know and I imagine that even not so newbies will hesitate...
>
> The toolbar not only displays icons but these icons have a legend which
> explain their purpose. How would a newbie (client) know that one or
> multiple items may be trashed, archived, checked-in, part of a batch, etc. ?
> People NEED to know what is available... It is essential for a good user
> experience.
>
> There was a proposal to dim the buttons. That would solve the issue.
> Another solution would be to create a new parameter in the template, and
> disable the hiding by default.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7592#issuecomment-126600555>.
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

infograf768 commented:
Obviously default to hide would not be useful to newbies, which is the purpose of the possible param.
Joomla has to move on, but not in a way that makes it a specialist-only CMS. Or is that what is looked for by some?

infograf768 commented:
@dgt41 
Also please look Clear Expired Cache where the toolbar button should always display : its div id is "toolbar-delete"

dgrammatiko committed:
change the opacity instead of totaly hiding the btns

dgt41 commented:
How about playing with the opacity of those buttons instead of totally hiding them?
I looks like this:
![screenshot 2015-07-31 14 18 21](https://cloud.githubusercontent.com/assets/3889375/9006373/566165d0-378f-11e5-80a8-fc1d4a833162.png)
![screenshot 2015-07-31 14 18 34](https://cloud.githubusercontent.com/assets/3889375/9006377/5bd684b4-378f-11e5-82a6-e8311902176b.png)

dgrammatiko committed:
fix finder purge

dgrammatiko committed:
fix clear expired cache

mbabker commented:
@infograf768 Generally I agree we shouldn't be making Joomla harder to use.  With that said though, Gmail uses a very similar toolbar behavior; is that a platform you've had a hard time learning what features are available in different conditions?

Proper UX testing on this would be the ultimate "right" answer in deciding whether it generally makes Joomla harder to learn/use or not.  But, we don't have that resource right now.  So this is ultimately a "best guess" or "I think this is cool" change without real data.

brianteeman commented:
Not just gmail. Many mainstream applications do this as well.
On 31 Jul 2015 14:23, "Michael Babker" <notifications@github.com> wrote:

> @infograf768 <https://github.com/infograf768> Generally I agree we
> shouldn't be making Joomla harder to use. With that said though, Gmail uses
> a very similar toolbar behavior; is that a platform you've had a hard time
> learning what features are available in different conditions?
>
> Proper UX testing on this would be the ultimate "right" answer in deciding
> whether it generally makes Joomla harder to learn/use or not. But, we don't
> have that resource right now. So this is ultimately a "best guess" or "I
> think this is cool" change without real data.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7592#issuecomment-126690313>.
>

brianteeman commented:
@dgt41 when they are opaque are they also disabled? <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7592">issues.joomla.org/joomla-cms/7592</a>.</sub>

dgt41 commented:
Unfortunately nope…

brianteeman commented:
Thought as much. So that's a big thumbs down from me as it doesn't resolve
anything

dgt41 commented:
I am also against this, but I thought it might be useful if people could try both options here, so there won’t be a biased decision!

infograf768 commented:
>So that's a big thumbs down from me as it doesn't resolve anything

Nothing needs to be solved as there is no problem with current implementation. No space gain for example. I consider this as a change for the sake of changing.

Why not go further ? Here is a possible solution:
![purejoomla_administration](https://cloud.githubusercontent.com/assets/869724/9020877/a2998f12-3826-11e5-822a-a7ca82df14c0.png)

(+ as I wrote above, the possible issues for 3pd)

brianteeman commented:
There is a problem in the current implementation - people complain that the
UI is cluttered and they find all the buttons overwhelming and we do
nothing to prevent a user selecting multiple records to edit but it only
edits one #7589



On 1 August 2015 at 07:28, infograf768 <notifications@github.com> wrote:

> So that's a big thumbs down from me as it doesn't resolve anything
>
> Nothing needs to be solved as there is no problem with current
> implementation. No space gain for example. I consider this as a change for
> the sake of changing.
>
> Why not go further ? Here is a possible solution:
> [image: purejoomla_administration]
> <https://cloud.githubusercontent.com/assets/869724/9020877/a2998f12-3826-11e5-822a-a7ca82df14c0.png>
>
> (+ as I wrote above, the possible issues for 3pd)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7592#issuecomment-126872194>.
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

infograf768 commented:
Editing only the first item in the list when selecting multiple records to edit has never been a real problem since Mambo times.  It should just be considered as a mistake by the user and Joomla deals with it nicely. 
As for the cluttered UI, it is an extremely subjective matter. As not all managers propose exactly the same functionalities and one may not use some on a daily basis, it forces to select an item to know what can be done or not...
Having to select an existing package in com_localise for example to know that you can (and sometimes you must when it is a master pack) duplicate it grace to the toolbar button is less than user friendly, and examples like this are numerous.

rbuelund commented:
Could we please get some more info statistics to enlighten this subject a bit more:

"people complain that the UI is cluttered" - how many people? - has there been a survey? I have not seen that
"Many mainstream applications do this as well" - please mention som others, so we can check them out.

Maybe this is a good idea and maybe not - but not just because GMail uses it - I would like to have the abillity to let som non experienced users try this out.

btoplak commented:
I welcome the idea of simplifying admin UI! Many rookie users dislike Joomla vs WP just because they get confused, and ultimately scared away with too many options to choose from. It's really time that Joomla gets ahead of this "issue".
But I also agree with @infograf768 that we should not clog existing Joomla users because of that.

So I suggest we should be more cunning on that subject. Let's do a fork of this backend template into two templates: "simple" and "power user". This way the switch between UI's is on the right place - by switching admin templates.
And, let's don't decide about the interface in the name of users. We could ask them to choose the UI (template) during installation, giving them a small substep to choose between "simple (new)" and "power user (standard)" UI.

btoplak commented:
Btw by choosing admin template during installation we are leaving all existing users on well known old admin template, so there won't be a problem too.

syntaxerror commented:
I have to agree with @infograf768, and salute the ideas by @btoplak - it really seems to be the best way dealing with it no matter what the end result is. But having them both greyed out and disabled,  still displaying the legend, by default - would also be nice.

Fedik commented:
I also liked idea in this pull request, but I also agree with @infograf768 ... click on each item to see what I can do with it will be really annoying ...
@btoplak idea about 2 templates just not real :wink:  

my suggestion,  do it "vise versa":
- by default display all buttons, as it is 
- when user select the item, just hide the buttons (or make small opacity) that cannot be used

btoplak commented:
@Fedik can you elaborate more why do you state 2 admin templates are not real?

Admininstrator Templates functionality is there for years, but not used a lot. The purpose should be to enable Joomlers an option to switch different backend UI, and we are talking exactly about this - backend UI changes. 
Adding extra Global Configuration options to achieve almost the same thing is just another step to make Joomla over-complicated system. And I think the whole purpose here is to make it simpler.

So, why not have:
1. "power user" UI/template: for existing Joomla users
2. "simplified" UI/template: for new/inexperienced users

And the choice of backend UI (aka admin template) would be given in installer, so experienced and new users can choose which UI suits them more.

Fedik commented:
@btoplak 2 template in the core means twice more time for support,
simple example: now we have `Isis` and `Hathor` .. peoples often just forget to check how their patch will work in `Hathor` .

so it is not real, from my point of view :wink:

justinherrin commented:
If a simple parameter or two in the admin template area could allow users to pick between seeing all the buttons or a simplified amount of buttons, I don't understand why two different admin Isis templates would be needed.

dgt41 commented:
@justinherrin that won’t be very hard to do, I will try to update to incorporate this. I would suggest to have default value for the existing users the old behavior and the new behavior for new installations. 
What do you think about this approach?

infograf768 commented:
@dgt41
Hiding the toolbar buttons is an issue for ALL users.
As I said above, one can't remember all Options possible as managers may propose different stuff.

I like the PR as it is with dimmed icons until an item is selected.
Default should be dimmed OR visible in ALL cases.
If desired, you could add a parameter to let people who like it this way to hide fully the buttons.

Forcing new install to not show the buttons would be counter-productive.

astridx commented:
I found another Problem. 
If there is only one Item in a List then the Box in the Head is selected automatically, if I select this one user. For example I have only one User in the UserManager and I select this one user.
 
So there is more than one selection and with the solution here the edit button is not active, also it should be active.

dgrammatiko committed:
implement suggestions

- Add switch on the template advanced tab
- Fix the bug for list containing only one item
- Restore hide/show instead of the dimmed version of the buttons

dgt41 commented:
@astridx Thanks for finding this! should be fixed now
@infograf768 @brianteeman @justinherrin @Fedik @btoplak I just introduce a switch on the advanced tab of isis for this feature. Default is the old known toolbar! Also the code is back to hide/show instead of dimming down the buttons. I hope this is acceptable now.

justinherrin commented:
@dgt41 don't forget the updated language file for the two new strings `TPL_ISIS_FIELD_ENHANCED_BUTTONS_LABEL` and `TPL_ISIS_FIELD_ENHANCED_BUTTONS_DESC`

Unless I am missing something...

dgrammatiko committed:
the lang file is also needed

peterlose commented on administrator/templates/isis/index.php:
```diff
@@ -96,6 +97,17 @@ function colorIsLight($color)
 	return $yiq >= 200;
 }
+
+if ($displayHeader)
```
Shouldn't this be if ($enhancedBtns) ?

dgrammatiko committed:
oops

infograf768 commented:
@dgt41 
The new field in the xml is not formatted as the others (spaces, indentation).

Otherwise, looks all is OK for me here.

zero-24 committed:
Update templateDetails.xml

Commit:
Merge pull request #20 from zero-24/patch-2

CS Thanks @zero-24

astridx commented:
This works correct for me.

But I would prefer the solution with the opaque buttons, in the case that this opaque buttons are disabled. 

I think creating a new parameter makes isis unnecessary more complex. And I do not expect that the unexperienced users will find this parameter very fast. And this solution should help them in the first instance.

In a further version of this patch I made a [PR where the opaque buttons are disabled](https://github.com/dgt41/joomla-cms/pull/19 "PR where the opaque buttons are disabled"), too:


Is this solution not possible? 
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7592">issues.joomla.org/joomla-cms/7592</a>.</sub>

dgt41 commented:
@astridx I like the idea not having one more option to choose in the template configuration. Also I am not biased on dimmed or hidden, although personally I think I like more the clean approach of hiding completely the buttons. But that comes with an extra option and there goes the infinite loop… So if people decide that dimming+disabling is better I will merge your PR, by the way thanks for that!

justinherrin commented:
I personally can see it both ways (show+hide vs. dimming+disabling), and I think both are great ideas. Is there any way we can do some user testing to a wider group of people rather than just us making the decision?

I hate to recommend this because I don't know how hard it would be to accomplish, but what if there was a 3rd setting in the `templateDetails.xml` file that allowed the user to pick which style to use?

```
<option value="2">Advanced: Dimming Option</option>
<option value="1">Advanced: Show+Hide Option</option>
<option value="0">Normal Isis Buttons</option>
```

Would something like that be real difficult to do?

astridx commented:
I understood the option like this: The show/hide option was just a compromise because the disabling was not implemented. 

But perhaps I misunderstood this. Anyway, I would prefer a solution without another option.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7592">issues.joomla.org/joomla-cms/7592</a>.</sub>

dgt41 commented:
I guess the UX team needs to make a decision on this one. I am gonna close it for now

PhilETaylor commented:
> I guess the UX team needs to make a decision on this one. I am gonna close it for now

A phrase I read FAR TOO OFTEN!
